bin/wasm-node/rust/src/ffi.rs

```diff
@@ -499,40 +499,52 @@ fn init(
         iter::once(super::ChainConfig {
             specification: chain_specs,
             database_content,
+            json_rpc_running: true,
         })
         .chain(
             parachain_specs
                 .into_iter()
                 .map(|specification| super::ChainConfig {
                     specification,
                     database_content: None,
+                    json_rpc_running: true,
                 }),
         ),
         max_log_level,
     ));
 }
 
+pub(crate) struct Request {
+    pub(crate) json_rpc_request: Box<[u8]>,
+    pub(crate) chain_index: usize,
+}
+
 lazy_static::lazy_static! {
-    static ref JSON_RPC_CHANNEL: (mpsc::UnboundedSender<Box<[u8]>>, futures::lock::Mutex<mpsc::UnboundedReceiver<Box<[u8]>>>) = {
+    static ref JSON_RPC_CHANNEL: (mpsc::UnboundedSender<Request>, futures::lock::Mutex<mpsc::UnboundedReceiver<Request>>) = {
         let (tx, rx) = mpsc::unbounded();
         (tx, futures::lock::Mutex::new(rx))
     };
 }
```

Comment:
Okay, so do we want multiple channels (1 per chain / 1 per json rpc service?) or a single channel that passes in a `chain_index`? If so, how should I rewrite `next_json_rpc` so that it waits on a request for a specific chain index and doesn't discard any requests that are for different chain index?

Reply:
I'd suggest to keep one channel, but instead of having the JSON-RPC service poll the channel directly, you spawn an additional new task common to all JSON-RPC services that is dedicated to polling it.
In other words, move [this piece of code](https://github.com/paritytech/smoldot/blob/ee4895be2a58399953aaca085c9a404361ae5a24/bin/wasm-node/rust/src/json_rpc_service.rs#L162-L201) outside of the JSON-RPC service, and make `handle_rpc` public.

Reply:
Okay, how would this task call `handle_rpc` on the clients? Should it just be passed in a `Vec`/`HashMap` of clients?

Reply:
Yes